[FLINK-12749] [docs] [examples] Initial Version of Flink Cluster Playground

knaufk opened the pull request:
## Purpose of this Change

Adds Flink Cluster Playground to Getting Started section of the documenation

## Brief change log

  - adds Flink Docker Playgrounds section to Getting Started guide
  - added additional example `ClickEventCount` to flink-examples
  - added first version of Flink Cluster Playground to documenation
  - [hotfix] upgrade Kafka Depdendency in flink-examples to the universal Kafka connector

## Verifying this change

* build documentation and checkout /getting-started/docker-playgrounds
* do a walkthrough of the complete playground
* For this you will need a a local Flink image `flink:1.9-scala_2.11` as the official docker images for 1.9 have naturally not been published yet. You can do this by building this branch locally, and replacing /opt/flink in `flink:1.8.1-scala_2.11`. For this checkout `docker-flink` and modify `https://github.com/docker-flink/docker-flink/blob/master/1.8/scala_2.11-debian/Dockerfile. 

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -23,5 +23,658 @@ specific language governing permissions and limitations
+
+### Flink WebUI
+
+The Flink WebUI is probably the most natural starting point to observe your Flink Cluster. It is
```
I find the wording here strange. How about: 

> The most natural starting point to observe your Flink Cluster is the Web UI exposed under http://localhost:8081.

knaufk replied:
Done.

sjwiesman commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/functions/ClickEventStatisticsCollector.java:
```diff
@@ -0,0 +1,45 @@
+
+		Long count = elements.iterator().next();
+
+		out.collect(new ClickEventStatistics(context.window().getStart(), context.window().getEnd(), page, count));
```
I originally started to write a comment that you should have `CountingAggregator#getResult` return a `ClickEventStatistics` object but now I see you are getting the start and stop times of the window. Maybe add a comment for new users who make the same mistake I did :)

sjwiesman commented:
@knaufk This all looks really polished, just a few nits

knaufk commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/ClickEventCount.java:
```diff
@@ -0,0 +1,105 @@
+ * <p> It can be run with or without checkpointing and with event time or processing time semantics.
+ * </p>
+ *
+ *
```
Good point. Done.

knaufk commented on docs/getting-started/docker-playgrounds/interactive_sql_playground.md:
```diff
@@ -0,0 +1,30 @@
+* This will be replaced by the TOC
+{:toc}
+
+This section will describe how to deploy and use a `docker-compose`-based playground centered around
```
Done.

knaufk commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/ClickEventGenerator.java:
```diff
@@ -0,0 +1,111 @@
+
+	private static final List<String> pages = Arrays.asList("/help", "/index", "/shop", "/jobs", "/about", "/news");
+
+	private static KafkaProducer<byte[], byte[]> producer;
```
Done.

knaufk commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/ClickEventGenerator.java:
```diff
@@ -0,0 +1,111 @@
+		return kafkaProps;
+	}
+
+	public static long nextTimestamp(String page) {
```
Arguably a better design, yes :) Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+variety, the fundamental building blocks of a Flink Cluster remain the same and similar
+operational principles apply.
+
+This docker compose-based playground will get you started with Apache Flink operations quickly and
```
Done.

knaufk commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/ClickEventCount.java:
```diff
@@ -0,0 +1,105 @@
+	public static void main(String[] args) throws Exception {
+		final ParameterTool params = ParameterTool.fromArgs(args);
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+  <b>Note</b>: The Apache Flink Docker images used for this playground are only available for
+  released versions of Apache Flink. Since you are currently looking at the latest SNAPSHOT
+  version of the documentation the branch referenced below will not exist. You can either change it 
+  manually or switch to the released version of the ocumentation via the release picker.
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+{% top %}
+
+## Setup
```
Yes, this is nicer.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+This indicates that the client container has successfully submitted the Flink Job ("Exit 0") and all 
+cluster components as well as the data generator are running ("Up").
+
+## Interaction with the Playground
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+## Interaction with the Playground
+
+There are many ways to interact with this playground and its components.
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+</p>
+{% endif %}
+
+In this section you will setup the playground locally on your machine and verify that the Job is 
```
That's more engaging. Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+{%  top %}
+
+## Tear Down
```
Fair point. I merged it into the the setup section.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+{%  top %}
+
+## Operational Plays
```
I went with "Time to Play".

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+### Observing Failure & Recovery
+
+Flink provides exactly-once processing guarantees under (partial) failure. In this playground you 
```
Well, you can not confirm that something is working by observing a correct behavior in one environment once, but maybe we don't need to be so modest here.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+Upgrading a Flink Job always involves two steps: First, the Flink Job is gracefully stopped with a
+[Savepoint]({{site.base_url}}/ops/state/savepoints.html). A Savepoint is a consistent snapshot of 
+the complete application state at well-defined, globally consistent point in time (similar to a 
+checkpoint). Second, the upgraded Flink Job is started from the Savepoint. Inthis context "upgrade" 
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+can mean different things including the following:
+
+* An upgrade to the configuration (incl. the parallelism of the Job)
+* An upgrade to the topology of the Job
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+}
+{% endhighlight %}
+
+**Request**
```
Why? To complicated?

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+The Flink Master exposes all (system & user) [metrics]({{ site.baseurl }}/monitoring/metrics.html)
+via its REST API. The endpoint depends on the scope of these metrics.
+
+#### Job Scope
```
Ok, done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+### Querying the Metrics of a Job
+
+The Flink Master exposes all (system & user) [metrics]({{ site.baseurl }}/monitoring/metrics.html)
```
Jepp.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+**Command**
+{% highlight bash %}
+ls -lia /tmp/flink-savepoints-directory
```
No, I have not. I'll try to find someone on Mac and Windows to go through the playground.

knaufk replied:
@sjwiesman has tested this on Mac, and @zentol on Windows.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+#### Step 1: Stopping the Job
+
+To gracefully stop the Job, you need to use the "stop" command of either the CLI or the REST API. 
+For this you will need the JobID of the Job, which is you can obtain by 
```
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+{% endhighlight %}
+
+Once the new TaskManager has registered itself with the Flink Master, the Job will start "RUNNING" 
+again. It will then quickly process the full backlog you can of input events from Kafka and produce output 
```
Done and done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+{% endhighlight %}
+</div>
+</div>
+
```
Done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+The Kafka Cluster consists of a Zookeeper server and a Kafka Broker.
+
+<img src="{{ site.baseurl }}/fig/flink-docker-playground.svg" alt="Flink Docker Playground"
```
Done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+When the playground is started a Flink Job called *Flink Event Count* will be submitted to the 
+Flink Master. Additionally, two Kafka Topics *input* and *output* are created.
+
+<img src="{{ site.baseurl }}/fig/click-event-count-example.svg" alt="Click Event Count Example"
```
Did this by using `java.util.Date` instead of `long` and added a comment that this was done for readability.

knaufk replied:
Done by using java.util.Date;

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+//input topic (1000 records/s)
+docker-compose exec kafka kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic input
+//output topic (6 records/min)
+docker-compose exec kafka kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic output
```
I reduced it to 15 seconds.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+[windows]({{ site.baseurl }}/dev/stream/operators/windows.html). The results are written to the 
+*output* topic. 
+
+There are six different `page`s and the **events are generated so that each window contains exactly 
```
Done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+operational principles apply.
+
+This docker compose-based playground will get you started with Apache Flink operations quickly and
+will briefly introduce you to the main components that make up a Flink Cluster.
```
Good point. Done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+### Flink REST API
+
+The [Flink REST API]({{ site.baseurl }}/monitoring/rest_api.html#api) is exposed via
```
I agree, done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+There are many ways to interact with this playground and its components.
+
+### Flink WebUI
```
I kinda like the structure here.

knaufk replied:
I would rather keep this structure.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+
+<img src="{{ site.baseurl }}/fig/playground-webui.png" alt="Playground Flink WebUI"
+class="offset" width="100%" />
+
```
Done.

knaufk replied:
Done.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+docker-compose kill taskmanager
+{% endhighlight %}
+
+After a few seconds, you will see in the Flink WebUI that the Job failed, and has been 
```
Same for me. I tried a few things already, but without success. Need to look it into further.

knaufk replied:
Same for me. I tried a few things already, but without luck so far. Need to look further into ít.

zentol replied:
shouldn't this be controller via `heartbeat.timeout`? By default it is set to 50 seconds.

knaufk replied:
I thought so, but neither `heartbeat.timeout` nor `akka.ask.timeout` anything related seemed to have any effect.

fhueske replied:
for me the following setting worked:

```
heartbeat.interval: 1000
heartbeat.timeout: 5000
```

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+</div>
+
+Once the Job is "RUNNING" again, you will see in the *output* Topic that now data was lost during 
+the upgrade: all windows are present with a count of exactly one thousand.
```
Done.

knaufk commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/clickeventcount/ClickEventCount.java:
```diff
@@ -0,0 +1,105 @@
+	public static final String CHECKPOINTING_OPTION = "checkpointing";
+	public static final String EVENT_TIME_OPTION = "event-time";
+
+	public static final Time WINDOW_SIZE = Time.of(60, TimeUnit.SECONDS);
```
Done.

knaufk commented:
Thanks a lot @sjwiesman and @fhueske for your reviews and valuable suggestions! I have addressed most of them by now and pushed a new version. 

For me the only blocker at this point is to test it on Windows and MacOs.

knaufk commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,680 @@
+the actual [Tasks]({{ site.baseurl }}/concepts/glossary.html#task) which make up a Flink Job. In 
+this playground you will start with a single TaskManager, but scale out to more TaskManagers later. 
+Additionally, this playground comes with a dedicated *client* container, which we use to submit the 
+Flink Job initially and to perform various operational tasks later on.
```
Done.

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,780 @@
+In this playground you will see to how to run Jobs on Flink and get a good impression of how to 
+manage and operate them. More specifically you will:
+
+* Run a Flink Job
```
For some reason I can't get this list and the TOC to render at the same time. It only shows up if comment out the table of contents. 

<img width="915" alt="Screen Shot 2019-07-31 at 4 16 03 PM" src="https://user-images.githubusercontent.com/1891970/62248832-ab3b9180-b3ae-11e9-88e8-1c98072105e1.png">

sjwiesman replied:
Yep, the only way they will both render is if I add another sentence in between. This looks like a bug in Jekyll. 
<img width="1189" alt="Screen Shot 2019-07-31 at 4 20 25 PM" src="https://user-images.githubusercontent.com/1891970/62249078-3e74c700-b3af-11e9-83dd-fb180ff10814.png">

sjwiesman replied:
I don't think its worth debugging Jekyll on this one, how about we just replace the bullet list with a sentence in the previous paragraph. 

> In this playground, you will learn how to manage and run Flink Jobs. You will see how to deploy and monitor an application, experience how Flink recovers from Job failure, and perform everyday operational tasks like upgrades and rescaling.

knaufk replied:
Done. Thanks for spotting this!

sjwiesman commented:
Other than the render issue +1 both for this and the PR in flink-playgrounds. I can confirm that this works on a mac.

knaufk commented:
@fhueske I have rebased and changed the target to `master`. From my point of view this is ready to be merged. The open comments are valuable suggestions and I think, we should implement them in the future. Right now, I'd rather get this into master, so that we might have a shot at still getting this into Flink 1.9.

fhueske commented:
Thanks for the update@knaufk! 
I'll have a final look and will try to merge it.
Fabian

fhueske commented:
OK, I'll do a few minor changes and will merge the PR.
Thanks for working on this @knaufk!

tillrohrmann commented:
Nice work @knaufk. I'm just wondering why does the `ClickEventCount` job needs to be part of the Flink distribution? Could it also work if it were part of the playgrounds docker images?

knaufk commented:
@tillrohrmann We decided against custom images for the playground to reduce the neccessary maintaince work. Instead we build on top of the existing `library/flink` images.

tillrohrmann commented:
Hmm, shouldn't it be somewhat straight forward to take a `library/flink` image and add an additional binary to it? I always thought that this onion layer abstraction is the very nice thing about Dockerfiles.

My concern is that we clutter Flink's distribution with examples which are not really useful if not used in combination with the playground docker images.

zjuwangg commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,774 @@
+commands and show how to validate that everything is running correctly.
+
+We assume that you have that you have [docker](https://docs.docker.com/) (1.12+) and
+[docker-compose](https://docs.docker.com/compose/) (2.1+) installed on your machine.
```
Sorry late to reply, I just noticed that here should be a docker-compose file version instead of a docker-compose version. Just a minor advice.
